python/examples/core/nevergrad_parallel_utils.py

```diff
@@ -39,71 +40,219 @@ def __init__(self, proposal, problem_instance):
+  assert benchmark_cpu_min <= compilation_cpu and compilation_cpu <= benchmark_cpu_max, \
+    f'compilation_cpu: {compilation_cpu} is not in range [{benchmark_cpu_min}, {benchmark_cpu_max})'
 
+  # Function called in the compilation thread.
+  def compile():
     # Construct the schedule and save the module in case we need to replay later.
     def schedule_and_save(module):
       scheduler.schedule(module, proposal)
       # TODO: save and report on error.
 
+    problem.compile_with_schedule_builder(
+        entry_point_name=scheduler.entry_point_name,
+        fun_to_benchmark_name=scheduler.fun_to_benchmark_name,
+        compile_time_problem_sizes_dict=scheduler.
+        build_compile_time_problem_sizes(),
+        schedule_builder=schedule_and_save)
+
+  throughputs_placeholder = []
+
+  # Function called in the benchmark thread.
+  def benchmark():
+    throughputs_placeholder.append(
+        problem.run(
+            n_iters=n_iters,
+            entry_point_name=scheduler.entry_point_name,
+            runtime_problem_sizes_dict=problem.compile_time_problem_sizes_dict))
+
+  try:
+    import threading
+
     f = io.StringIO()
     with redirect_stdout(f):
-      problem.compile_with_schedule_builder(
-          entry_point_name=scheduler.entry_point_name,
-          fun_to_benchmark_name=scheduler.fun_to_benchmark_name,
-          compile_time_problem_sizes_dict= \
-            scheduler.build_compile_time_problem_sizes(),
-          schedule_builder=schedule_and_save)
-
-      throughputs = problem.run(
-          n_iters=n_iters,
-          entry_point_name=scheduler.entry_point_name,
-          runtime_problem_sizes_dict=problem.compile_time_problem_sizes_dict)
+      # Pin compile part of the process to the unique compilation_cpu.
+      os.system(
+          f'taskset -p -c {compilation_cpu} {os.getpid()} > /dev/null 2>&1')
+
+      # Acquire, spawn a new thread, start and join with timeout, release.
+      lock.acquire_read()
+      t = threading.Thread(target=compile)
+      t.start()
+      t.join(timeout_compile)
+      lock.release()
+      if t.is_alive():
+        os.system('kill %d' % os.getpid())
+        return
+
+      # Pin benchmark part of the process to the cpu_min-cpu_max range to allow
+      # parallel evaluations without prohibitive interferences.
+      os.system(
+          f'taskset -p -c {benchmark_cpu_min}-{benchmark_cpu_max} {os.getpid()} > /dev/null 2>&1'
+      )
+
+      # Acquire, spawn a new thread, start and join with timeout, release.
+      lock.acquire_write()
+      t = threading.Thread(target=benchmark)
+      t.start()
+      t.join(timeout_benchmark)
+      lock.release()
+      if t.is_alive():
+        os.system('kill %d' % os.getpid())
+        return
 
     # TODO: redirect to a file if we want this information.
     f.flush()
 
-    ipc_dict['result'] = IPCState(success=True, throughputs=throughputs)
+    ipc_dict['result'] = IPCState(success=True,
```

Comment:
I think this must also be set in case of a failure? Before killing the process

Reply:
NGMPEvaluation checks whether result is set.
Doing it here will be tricky because the process is not joined and information does not pass easily.